Malformed XML leads to premature node exit in some cases

Unfortunately, I was not able to isolate this repro any further so here's how to get to the same state in dotnet/sdk where I discovered this:
- mkdir repro
- cd repro
- git clone https://github.com/nguerrera/sdk
- cd sdk
- git checkout msbuild-repro
- build

**Expected result**
 An error indicating that the targets file is malformed. 

See https://github.com/nguerrera/sdk/commit/f752fe35e0c18f96632d597a4a1e3983d79b0b8f#commitcomment-19567654

**Actual result**

```
MSBUILD : error MSB4166: Child node "5" exited prematurely. Shutting down. Diagnostic 
information may be found in files in the temporary files directory named MSBuild_*.failure.txt.

UNHANDLED EXCEPTIONS FROM PROCESS 11084:
=====================
10/25/2016 1:10:55 PM
System.IO.IOException: Pipe is broken.
   at System.IO.Pipes.PipeStream.WriteCore(Byte[] buffer, Int32 offset, Int32 count)
   at System.IO.Pipes.PipeStream.Write(Byte[] buffer, Int32 offset, Int32 count)
   at Microsoft.Build.BackEnd.NodeEndpointOutOfProcBase.PacketPumpProc()
===================
```

@raines @jeffkl

--- a/src/Shared/NodePacketTranslator.cs
+++ b/src/Shared/NodePacketTranslator.cs
@@ -18,6 +18,7 @@ using System.Runtime.Serialization.Formatters.Binary;
 using Microsoft.Build.Collections;
 using Microsoft.Build.Execution;
 using Microsoft.Build.Framework;
+using Microsoft.Build.Internal;
 using Microsoft.Build.Shared;
 using System.Globalization;
 using System.Reflection;
@@ -1396,35 +1397,44 @@ namespace Microsoft.Build.BackEnd
 
             public static Exception ToException(SerializedException serializedException)
             {
-                var typeLoader = new TypeLoader((t, o) => true);
-
-                LoadedType loadedExceptionType = typeLoader.Load(serializedException.TypeFullName, serializedException.ExceptionAssembly);
-                Type exceptionType = loadedExceptionType.Type;
-
                 Exception innerException = null;
                 if (serializedException.InnerException != null)
                 {
                     innerException = ToException(serializedException.InnerException);
                 }
 
-                Type[] parameterTypes;
-                object[] constructorParameters;
-                if (innerException == null)
+                ConstructorInfo constructor = null;
+                object[] constructorParameters = null;
+
+                try
                 {
-                    parameterTypes = new[] { typeof(string) };
-                    constructorParameters = new object[] { serializedException.Message };
+                    var typeLoader = new TypeLoader((t, o) => true);
+
+                    LoadedType loadedExceptionType = typeLoader.Load(serializedException.TypeFullName, serializedException.ExceptionAssembly);
+                    Type exceptionType = loadedExceptionType.Type;
+
+                    Type[] parameterTypes;
+                    if (innerException == null)
+                    {
+                        parameterTypes = new[] { typeof(string) };
+                        constructorParameters = new object[] { serializedException.Message };
+                    }
+                    else
+                    {
+                        parameterTypes = new[] { typeof(string), typeof(Exception) };
+                        constructorParameters = new object[] { serializedException.Message, innerException };
+                    }
+
+                    constructor = exceptionType.GetConstructor(parameterTypes);
                 }
-                else
+                catch (FileLoadException e)
                 {
-                    parameterTypes = new[] { typeof(string), typeof(Exception) };
-                    constructorParameters = new object[] { serializedException.Message, innerException };
+                    CommunicationsUtilities.Trace($"Exception while attempting to deserialize an exception of type \"{serializedException.TypeFullName}\". Could not load from \"{serializedException.ExceptionAssembly.AssemblyLocation}\": {e}");
                 }
 
-                ConstructorInfo constructor = exceptionType.GetConstructor(parameterTypes);
                 if (constructor == null)
                 {
-                    //  Couldn't find appropriate constructor.  Fall back to creating an exception
-                    //  that will look the same as the original one when ToString() is called
+                    CommunicationsUtilities.Trace($"Could not find a constructor to deserialize an exception of type \"{serializedException.TypeFullName}\". Falling back to an exception that will look the same.");
                     return new FormattedException(serializedException.ExceptionToString);
                 }
 

